build2cmake/src/config/v2.rs

```diff
@@ -0,0 +1,195 @@
+#[serde(deny_unknown_fields)]
+pub struct Build {
+    pub general: General,
+    pub torch: Option<Torch>,
+
+    #[serde(rename = "kernel", default)]
+    pub kernels: HashMap<String, Kernel>,
+}
+
+impl Build {
+    pub fn has_kernel_with_backend(&self, backend: &Backend) -> bool {
+        self.kernels
+            .values()
+            .any(|kernel| kernel.backend == *backend)
+    }
+
+    pub fn backends(&self) -> BTreeSet<Backend> {
+        self.kernels.values().map(|kernel| kernel.backend).collect()
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+#[serde(deny_unknown_fields, rename_all = "kebab-case")]
+pub struct General {
+    pub name: String,
+    #[serde(default)]
+    pub universal: bool,
+}
+
+#[derive(Debug, Deserialize, Clone, Serialize)]
+#[serde(deny_unknown_fields)]
+pub struct Torch {
+    pub include: Option<Vec<String>>,
+    pub pyext: Option<Vec<String>>,
+
+    #[serde(default)]
+    pub src: Vec<PathBuf>,
+}
+
+impl Torch {
+    pub fn data_globs(&self) -> Option<Vec<String>> {
+        match self.pyext.as_ref() {
+            Some(exts) => {
+                let globs = exts
+                    .iter()
+                    .filter(|&ext| ext != "py" && ext != "pyi")
+                    .map(|ext| format!("\"**/*.{}\"", ext))
+                    .collect_vec();
+                if globs.is_empty() {
+                    None
+                } else {
+                    Some(globs)
+                }
+            }
+
+            None => None,
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+#[serde(deny_unknown_fields, rename_all = "kebab-case")]
+pub struct Kernel {
+    pub backend: Backend,
+    pub cuda_capabilities: Option<Vec<String>>,
+    pub rocm_archs: Option<Vec<String>>,
+    pub depends: Vec<Dependencies>,
+    pub include: Option<Vec<String>>,
+    pub src: Vec<String>,
+}
+
+#[derive(Clone, Copy, Debug, Deserialize, Eq, Ord, PartialEq, PartialOrd, Serialize)]
+#[serde(deny_unknown_fields, rename_all = "kebab-case")]
+pub enum Backend {
+    Cuda,
+    Rocm,
+}
+
+impl Display for Backend {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Backend::Cuda => write!(f, "cuda"),
+            Backend::Rocm => write!(f, "rocm"),
+        }
+    }
+}
+
+impl FromStr for Backend {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.to_lowercase().as_str() {
+            "cuda" => Ok(Backend::Cuda),
+            "rocm" => Ok(Backend::Rocm),
+            _ => Err(format!("Unknown backend: {}", s)),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
+#[non_exhaustive]
+#[serde(rename_all = "lowercase")]
+pub enum Dependencies {
+    #[serde[rename = "cutlass_2_10"]]
+    Cutlass2_10,
+    #[serde[rename = "cutlass_3_5"]]
+    Cutlass3_5,
+    #[serde[rename = "cutlass_3_6"]]
+    Cutlass3_6,
+    #[serde[rename = "cutlass_3_8"]]
```

Comment:
unsure about `#[serde[rename` is this different than `#[serde(rename`?

Reply:
Not sure, surprising that it works. I have changed it to use parentheses.